prov/sm2/src/sm2_coordination.c

```diff
@@ -0,0 +1,574 @@
+/*
+ * Copyright (c) Amazon.com, Inc. or its affiliates. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "sm2_coordination.h"
+#include "sm2.h"
+#include "sm2_atom.h"
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#define NEXT_MULTIPLE_OF(x, mod) x % mod ? ((x / mod) + 1) * mod : x
+#define ZOMBIE_ALLOCATION_NAME	 "ZOMBIE"
+#define SM2_COORDINATION_DIR	 "/dev/shm"
+#define SM2_COORDINATION_FILE	 SM2_COORDINATION_DIR "/fi_sm2_mmaps"
+#define SM2_STARTUP_MAX_TRIES	 1000
+
+static void sm2_file_attempt_shrink(struct sm2_mmap *map);
+int sm2_entry_lookup(const char *name, struct sm2_mmap *map);
+static void sm2_file_extend_for_region(struct sm2_mmap *map,
+				       int last_valid_entry);
+
+static inline bool pid_lives(int pid)
+{
+	int err = kill(pid, 0);
```

Comment:
what!?!?!?
Interesting choice of function name given it does exactly the opposite, and does something it shouldn't do anyway.

Reply:
https://stackoverflow.com/questions/9152979/check-if-process-exists-given-its-pid

There was a comment here but I was advised to remove it.